### PR TITLE
[FIX] Preserve common_range in kmer_hash_view

### PR DESCRIPTION
**Pull request by eseiler**
Resolves https://github.com/seqan/seqan3/issues/1988
Closes https://github.com/seqan/product_backlog/issues/161

### CODE REVIEW

File: include/seqan3/range/views/kmer_hash.hpp [hunk 1]
```diff
@@ -153,7 +153,17 @@ class kmer_hash_view : public std::ranges::view_interface<kmer_hash_view<urng_t>
      */
     auto end() noexcept
     {
-        return std::ranges::end(urange);
+        if constexpr (std::ranges::common_range<urng_t>)
```
Comment (smehringer) [inline include/seqan3/range/views/kmer_hash.hpp:156]: Hm.. Why don't we always return a shape iterator then? Does the urange sentinel have any advantages?

File: include/seqan3/range/views/kmer_hash.hpp [hunk 2]
```diff
@@ -171,7 +191,17 @@ class kmer_hash_view : public std::ranges::view_interface<kmer_hash_view<urng_t>
...
+        {
+            return shape_iterator<urng_t const>{std::next(std::ranges::begin(urange),
+                                                          std::ranges::distance(urange) - shape_.size() + 1),
+                                                std::ranges::end(urange),
+                                                shape_};
+        }
+        else
+        {
+            return end();
+        }
```
Comment (smehringer) [inline include/seqan3/range/views/kmer_hash.hpp:204]: This shouldn't be necessary since `end()` already has the if constexpr?

File: include/seqan3/range/views/kmer_hash.hpp [hunk 3]
```diff
@@ -294,7 +324,7 @@ class kmer_hash_view<urng_t>::shape_iterator
     {
         assert(std::ranges::size(shape_) > 0);
 
-        if (shape_.size() <= std::ranges::distance(text_left, text_right))
+        if (shape_.size() <= std::ranges::distance(text_left, text_right) + 1)
```
Comment (smehringer) [inline include/seqan3/range/views/kmer_hash.hpp:327]: Why this change?

File: include/seqan3/range/views/kmer_hash.hpp [hunk 4]
```diff
@@ -511,26 +511,15 @@ class kmer_hash_view<urng_t>::shape_iterator
     }
 
     /*!\anchor shape_iterator_operator-difference
-     * \brief Return offset between this and remote iterator's position.
-     * \attention This function is only available if `it_t` models std::random_access_iterator.
-     */
-    difference_type operator-(shape_iterator const & lhs) const noexcept
```
Comment (smehringer) [inline include/seqan3/range/views/kmer_hash.hpp:487]: Didn't you just introduce this operator in the first commit ?

File: include/seqan3/range/views/kmer_hash.hpp (hunk 1 above)
Comment (eseiler) [reply include/seqan3/range/views/kmer_hash.hpp:156]: Yes, apparently we can just guarantee common_range; I did something before where it did not work, but now it works.

File: include/seqan3/range/views/kmer_hash.hpp (hunk 2 above)
Comment (eseiler) [reply include/seqan3/range/views/kmer_hash.hpp:204]: I can already remote the cbegin and cend in this PR.

File: include/seqan3/range/views/kmer_hash.hpp (hunk 3 above)
Comment (eseiler) [reply include/seqan3/range/views/kmer_hash.hpp:327]: Basically:

```
        // shape size = 3
        // Text:     1 2 3 4 5 6 7 8 9
        // it_start: ^
        // it_end:       ^
        // distance(it_start, it_end) = 2
```

But I'm still investigating other stuff.

File: include/seqan3/range/views/kmer_hash.hpp (hunk 4 above)
Comment (eseiler) [reply include/seqan3/range/views/kmer_hash.hpp:487]: In the first commit I introduced the friend `operator-`, here I remove the member `operator-`

**Comment by eseiler** [comment]
As it turns out, the iterator semantics were wrong, i.e. the end() was not pointing to the end (or rather, one past the end).
I still need to simplify the code and with forward_list there is still a ASAN error.

File: include/seqan3/range/views/kmer_hash.hpp (hunk 1 above)
Comment (eseiler) [reply include/seqan3/range/views/kmer_hash.hpp:148]: No, we can't, because in order to retain the semantics of an end_iterator, we need to be able to assign the end of the underlying range to the `text_right` iterator of the `shape_iterator`. This only works if both have a common type, i.e. the range is a common range.

Comment (smehringer) [reply include/seqan3/range/views/kmer_hash.hpp:148]: Thanks for explaining!

File: include/seqan3/range/views/kmer_hash.hpp [hunk 5]
```diff
@@ -290,16 +281,66 @@ class kmer_hash_view<urng_t>::shape_iterator
...
+        // text_left: ^
+        // text_right:    ^
+        // distance(text_left, text_right) = 2
+        if (shape_.size() <= std::ranges::distance(text_left, text_right) + 1)
+        {
+            roll_factor = pow(sigma, static_cast<size_t>(std::ranges::size(shape_) - 1));
             hash_full();
         }
+
+        text_right = it_end;
```
Comment (smehringer) [inline include/seqan3/range/views/kmer_hash.hpp:343]: Does `hash_full` change the text_right, or why are you doing the assignment here again?

Comment (eseiler) [reply include/seqan3/range/views/kmer_hash.hpp:343]: Yes, it changes the iterators and text_right would point to the last element of the text, instead to the end

File: include/seqan3/range/views/kmer_hash.hpp [hunk 6]
```diff
@@ -290,16 +281,66 @@ class kmer_hash_view<urng_t>::shape_iterator
...
+    *
+    * Linear in size of shape.
+    */
+    shape_iterator(it_t it_start, sentinel_t it_end, shape s_, bool SEQAN3_DOXYGEN_ONLY(is_end)) : shape_{s_}
+    {
+        assert(std::ranges::size(shape_) > 0);
+
+        auto urange_size = std::ranges::distance(it_start, it_end);
+        auto step = (shape_.size() + 1 > urange_size) ? 0 : urange_size - shape_.size() + 1;
+        text_left = std::ranges::next(it_start, step, it_end);
```
Comment (rrahn) [inline include/seqan3/range/views/kmer_hash.hpp:329]: Mhmm, since we do this anyway, would it not be sufficient to use this in the end function and call the old constructor with text_it put to the respective position?

Comment (eseiler) [reply include/seqan3/range/views/kmer_hash.hpp:329]: If I got it correct, no, because then the right text iterator would point one before the end instead to the end

Comment (rrahn) [reply include/seqan3/range/views/kmer_hash.hpp:329]: Ok, so if I get it right, then the shape_iterator stores the begin and the end of the current infix over the text to compute the hash. That would mean text_lext points to the first valid character in the shape and text_right to the one behind the last one of the shape. Which is ok, because we ensure that after the last increment (when we reached the end) the text_right points to the regular end of the underlying range (in case it is a common range). In the other case we compare the text_right with the sentinel. So why does it point to one before the end?

Comment (eseiler) [reply include/seqan3/range/views/kmer_hash.hpp:329]: I have to look at it again, it was quite confusing to make it point to where it should.
